src/util/src/algebra.rs

```diff
@@ -0,0 +1,582 @@
+    )]
+    g: BigUint,
+    /// Group order `q`. Constructors computed it from the other fields.
+    #[serde(
+        serialize_with = "crate::biguint_serde::biguint_serialize",
+        deserialize_with = "crate::biguint_serde::biguint_deserialize"
+    )]
+    q: BigUint,
+}
+
+impl GroupElement {
+    /// Multiplies the group element with another group element.
+    ///
+    /// That is the function computes `(self * other) mod p` where `p` is the group modulus.
+    pub fn mul(&self, other: &GroupElement, group: &Group) -> GroupElement {
+        GroupElement((&self.0 * &other.0) % &group.p)
+    }
+
+    /// Computes the (multiplicative) inverse of a group element.
+    ///
+    /// For valid group elements this function will always return some value.
+    pub fn inv(&self, group: &Group) -> Option<Self> {
+        mod_inverse(&self.0, &group.p).map(GroupElement)
+    }
+
+    /// Performs modular exponentiation of the group element with a given integer exponent.
+    pub fn pow<T>(&self, exponent: T, group: &Group) -> GroupElement
+    where
+        BigUint: From<T>,
+    {
+        let x = BigUint::from(exponent);
+        GroupElement(self.0.modpow(&x, &group.p))
+    }
+
+    /// Performs modular exponentiation of the group element with a given field element.
+    ///
+    /// This defines an action of the field over the group.
+    pub fn exp(&self, exponent: &FieldElement, group: &Group) -> GroupElement {
+        GroupElement(self.0.modpow(&exponent.0, &group.p))
+    }
+
+    /// Checks if the element is a valid member of the given group.
+    ///
+    /// This method return true iff `0 <= self < p` and `self^q % p == 1` where `p` is the group modulus and `q` the group order.
+    pub fn is_valid(&self, group: &Group) -> bool {
+        // It is enough to check the upper bound as self.0 is unsigned.
+        let elem_less_than_p = self.0 < group.p;
+        let elem_has_order_q = self.0.modpow(&group.q, &group.p).is_one();
+        elem_less_than_p && elem_has_order_q
+    }
+
+    /// Returns the left padded big-endian encoding of the group element.
+    ///
+    /// The encoding follows Section 5.1.1 in the specs.
+    pub fn to_be_bytes_left_pad(&self, group: &Group) -> Vec<u8> {
+        to_be_bytes_left_pad(&self.0, group.l_p())
+    }
+
+    pub fn to_biguint(&self) -> BigUint {
+        self.0.clone()
+    }
+}
+
+impl Group {
+    /// Constructs a new multiplicative integer group `Z_p^r`.
+    ///
+    /// The arguments are
+    /// - `modulus` - the modulus `p`
+    /// - `cofactor`- the cofactor `r`
+    /// - `generator` - a generator `g`
+    ///
+    /// This function checks that the group is valid according to [`Group::is_valid`].
+    /// Checking the validity of inputs is expensive.
+    /// A group should therefore be constructed once and then reused as much as possible.
+    ///
+    /// Alternatively, one can use fixed, *trusted/tested* parameters with [`Group::new_unchecked`].
+    pub fn new(
```

Comment:
With this function, I can say
`new(19, 5, 7, ...)` and get
`Some(Group { p: 19, r: 5, g: 7, q: 3 })` indicating that the group is valid although
19-1 = 18 **≠** 15 = 5*3. 

Would it maybe be easier take in `q` instead of `r` and not have `r` in the `Group` struct? And then let `is_valid` check what is needed.

Reply:
I think if can keep `r`, but there should be a check in `is_valid()` that `p-1 = qr`

Reply:
I have added the check to `is_valid()` and created a test out of your example.

Reply:
After discussion with @eb-concordium decided to remove the cofactor `r` from the `Group` struct as it is only need in `is_valid()`. This also simplifies the structure of `is_valid()`.